app/vminsert/netstorage/netstorage.go

```diff
@@ -568,8 +611,17 @@ func initStorageNodes(addrs []string, hashSeed uint64) *storageNodesBucket {
 			}
 			return 0
 		})
+		var nodeID uint64
+		if *usePersistentStorageNodeID {
+			nodeID = sn.getID()
+		} else {
+			nodeID = xxhash.Sum64String(addr)
+		}
+
+		nodeIDs = append(nodeIDs, nodeID)
 		sns = append(sns, sn)
 	}
+	nodesHash := newConsistentHash(nodeIDs, hashSeed)
```

Comment:
For now, it looks like if all vmstorages were down at the moment of vminsert start, all metric flow will be sent to the first node at nodeIDs array, since all nodeis == 0.

Reply:
It's a complicated problem. Yes, it must be dynamically updated to be consistent for all vminsert nodes.

 Most probably, storageNodes with id == 0 could be placed into routing with broken state.

Reply:
> I've added a fallback in [bb8b5db](https://github.com/VictoriaMetrics/VictoriaMetrics/commit/bb8b5db0fefc54726a31c5386e61c4c04a6bfb5b), do you think it will also be useful to implement a way for consistent has rebuild in order to get the real node IDs once vmstorage nodes will be available?

@zekker6 
Will generating a nodeId as a fallback cause other issues? The generated nodeId is completely different from the actual nodeId of the vmstorage node. In a consistent hash, this corresponds to a completely different node, which can lead to different time series shard distributions.

Additionally, if a vmstorage node fails after running a period of time, vminsert might still know the nodeId of that vmstorage. If some vminsert nodes restart, will this cause inconsistency in the nodeId view among different vminsert nodes?

Reply:
@shichanglin5 We've already discussed this and came up with exactly the same issues as you've described. Becase of this, the behaviour will be changed(currently WIP) to the following:
- if using persistent storage node IDs is enabled and storage node is not available on start it will not be a part of nodes pool
- once node will become reachable it will be added in pool with its persistent ID

This allows to ensure that persistent IDs will always be in sync and will minimize rerouting due to changes in consistent hash.

Reply:
nice